pkg/alertmanager/alertmanager_metrics.go

```diff
@@ -154,6 +154,10 @@ func (m *alertmanagerMetrics) addUserRegistry(user string, reg *prometheus.Regis
 	m.regs.AddUserRegistry(user, reg)
 }
 
+func (m *alertmanagerMetrics) removeUserRegistry(user string) {
+	m.regs.RemoveUserRegistry(user, true)
```

Comment:
"hard" remove is only possible under certain conditions:
1) all metrics are grouped "per user"
2) metrics that are not "per user" are only gauges

Unfortunately in this case there are multiple counters (eg. `silencesQueriesTotal`), and hard-remove will break them.

Reply:
Thanks for breaking that down - I went through a rabbit hole trying to understand when soft vs hard was possible by analysing `TestMetricsArePerUser` in the ruler but couldn't really make sense of it. I then noticed that most of the alertmanager metrics (even some I think should be) are not per user and that this was causing a problem with hard deletion.

In the world where we're going with a RF=3 as a default, I'm thinking that most of these metrics most be per user to make them useful.

Reply:
If you go with hard delete, then I'd suggest adding a test that verifies the two conditions I wrote above. See `TestMetricsArePerUser` test in ruler that verifies (1).